addons/yatl/chat/chat.gd

```diff
@@ -1,163 +1,171 @@
+
+# URL constants
+
 const CHAT_WSS_URL: String = "wss://irc-ws.chat.twitch.tv:443"
 
 
-#Signals
-signal chat_connected(yatl_instance)
-signal chat_login(yatl_instance, success) # success: bool
-signal chat_disconnected(yatl_instance, clean) # clean: bool
-signal chat_connection_failed(yatl_instance)
-signal chat_joined(yatl_instance, channel) # channel: String
-signal chat_message_received(yatl_instance, chat_message) # chat_message: ChatMessage
-signal chat_whisper_received(yatl_instance, chat_message) # chat_message: ChatMessage
-signal chat_raw_data_received(yatl_instance, raw_data) # raw_data: String
-signal chat_ping_received(yatl_instance)
+# Public variables
+
+var reconnect: bool = false # Whether or not to try and reconnect automatically
+var message_queue_timeout: float = 0.33 # how long to wait after sending each message
+
+
+# Private variables
 
+var __pal: __PAL
 
-#Private variables
-var __socket: WebSocketClient = WebSocketClient.new()
-var __channel_list: Array = []
+var __client: WebSocketClient
+var __socket_connection: __WebSocketConnection
 
+var __current_channel: String
 var __username_pattern: String = ":([\\w]+)!"
 var __username_regex = RegEx.new()
 
+var __message_queue: Array = []
+var __time_elapsed: float = 0.0
 
-#Lifecycle methods
-func _init() -> void:
-	__socket.verify_ssl = true
 
+# Lifecycle methods
+
+func _init(pal: __PAL) -> void:
+	__pal = pal
+	
 	__username_regex.compile(__username_pattern)
 
-	__socket.connect("connection_closed", self, "__socket_connection_closed")
-	__socket.connect("connection_established", self, "__socket_connection_established")
-	__socket.connect("connection_error", self, "__socket_connection_error")
-	__socket.connect("data_received", self, "__socket_data_received")
-	__socket.connect("server_close_request", self, "__socket_server_close_request")
 
+func _process(delta: float) -> void:
+	if !__socket_connection && !__socket_connection.connected:
+		return
+	
+	if !__message_queue.empty() && __time_elapsed >= message_queue_timeout:
+		__client_put_message(__message_queue.pop_front())
+		__time_elapsed = 0.0
+	else:
+		__time_elapsed += delta
 
-func _process(_delta: float) -> void:
-	if __socket.get_connection_status() != __socket.CONNECTION_DISCONNECTED:
-		__socket.poll()
 
+# Public methods
 
-#Public methods - chat
 func chat_connect() -> void:
-	print("chat connect called")
-	if __socket.connect_to_url(CHAT_WSS_URL) != OK:
-		emit_signal("chat_connection_failed", self)
+	if __socket_connection && __socket_connection.connected:
+		return
+
+	__socket_connection = __pal.establish_connection(CHAT_WSS_URL)
+	yield(__socket_connection, "completed")
+	if __socket_connection.connected:
+		__socket_connection.connect("disconnected", self, "__on_socket_connection_disconnected")
+		__socket_connection.connect("data_received", self, "__on_socket_connection_data_received")
+		__client = __socket_connection.client
+		emit_signal("connected")
+	else:
+		emit_signal("connection_failed")
```

Comment:
So the reason I made it a single signal is because anyone using `yield` have to do gymnastics to be able to wait for a result. I'm wondering if we should keep that pattern or not.

Reply:
That's fair, I didn't think of that. We can unify these signals, and maybe even lump in `disconnected`.